V7 support for request batching

fselmo opened the pull request:
### What was wrong?

Closes #832

### How was it fixed?

Add batch request support to the `RequestManager` via a context manager. Expose this as a high-level API on the `Web3` / `AsyncWeb3` class itself (`w3.batch_requests()`).

### Todo:

- [x] Handle / skip `EthereumTesterProvider` test case
- [x] Handle IPC & WS provider batch request cases
- [x] Handle persistent connection provider cases
- [x] Clean up commit history
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)
- [ ] Add or update documentation related to these changes -- separate PR ?

#### Cute Animal Picture

<img width="723" alt="Screenshot 2024-04-30 at 15 38 09" src="https://github.com/ethereum/web3.py/assets/3532824/f0833b5a-c3d7-4b7f-a742-ce3951d6fe2b">

dmkulazhenko committed:
Add basic support of batch requests to `AsyncHTTPProvider`

dmkulazhenko committed:
Fix typing for Python<3.9

Built-in types are not generics

fselmo committed:
Initial commit for batch requests support

- Implement some basic batching support for sync and async, running each
  request through request processing and response processing.
- Achieve this through context manager that prevents request sending while
  batching is in progress and re-allows when batching is done.
- Refactor batch middleware logic into Web3Middleware class
- Fix up the sync batch method to work by using the ``w3._is_batching``
  flag correctly.
- Refactor the batching logic in the manager to be a bit more readable
  and more performant - even if ever so slightly.
- Interestingly, using ``asyncio.gather()`` with x requests is still
  faster than batching the same amount of requests into a single
  batch request. More research into performance improvements is
  needed.
- Iterate on performance improvement ideas for batching
- Add basic tests for batch requests; minor typing & refactor

fselmo committed:
Batch requests async improvements; implement ``add_mapping()``:

- When adding async requests to a batch for a batch request, handle the
  awaiting of extracting the request information in the manager instead
  of adding that overhead to the batch API.
- Implement ``add_mapping()`` for batch requests

fselmo committed:
Add batch support for contract calls

fselmo committed:
Add batch support to remaining providers:

- Legacy websocket
- IPC
- Persistent connection providers (async_ipc, websocket)

fselmo committed:
Request information cache updates for request batching [persistent conn]

fselmo committed:
Move batching to web3 module; Re-structure Web3Module tests:

- Move request batching to a higher level public API on the ``web3`` module;
  make the implementation methods on the manager private.
- Create a proper ``AsyncWeb3ModuleTest`` class that can test differences
  between synchronous and asynchronous methods on the module that now exist.
  This gets rid of the parameterized tests that previously tested the
  static methods on both the ``Web3`` and the ``AsyncWeb3`` classes.
- Tighten up typing related to batch requests; add unsupported methods for batching with tests

fselmo committed:
Move provider type validation to ``_batch_requests`` method.

fselmo commented:
This is good for an initial review. It's still missing documentation but that may be best in a separate PR? Thoughts there? Usage examples can be seen in the tests for now.

fselmo committed:
Return responses sorted by response ids in each provider case

fselmo committed:
Refactor batch logic to allow for object assignment as per #832

pacrob commented on tests/integration/go_ethereum/test_goethereum_http.py:
```diff
@@ -70,7 +71,7 @@ def w3(geth_process, endpoint_uri):
 
 
-class TestGoEthereumTest(GoEthereumTest):
+class TestGoEthereumWeb3ModuleTest(GoEthereumWeb3ModuleTest):
```
There's a lot of this renaming in the integration tests. Is it just cleaning up for clarity, or more relevant in a way I'm not seeing?

fselmo replied:
The biggest change is making a proper `AsyncWeb3ModuleTest` and then yeah the renaming is for clarity. The old `Web3ModuleTest` was checking all the static utility methods for conversions, etc. Since we have some actual async methods on the `AsyncWeb3` class directly now, this PR adds a separation of logic between sync and async and gets rid of the `@pytest.mark.parametrize` that cycled between `Web3` and `AsyncWeb3`, opting for two separate test classes instead.

fselmo commented on web3/_utils/batching.py:
```diff
@@ -0,0 +1,171 @@
+}
+
+
+class BatchRequestContextManager(Generic[TFunc]):
```
Now that it's no longer just a context manager, e.g.:

```py
batch = w3.batch_requests()
batch.add(...)
batch.add(...)
result1, result2 = batch.execute()
```

... this class should probably be renamed to a more general name like `RequestBatcher`. We already have `RequestManager` and `RequestProcessor` classes. `RequestBatcher` seems like a fitting enough name.

---
edit: done in latest commit

fselmo committed:
Some changes to batcher API; More tests for robustness:

- Generalizaing name change: ``BatchRequestContextManager`` -> ``RequestBatcher``
  since it is no longer just a context manager.
- Add API to batcher for clearing (``clear``) and cancelling (``cancel``) the batch.
- Add more tests around these changes.

fselmo committed:
Make sure we sort list responses before caching

kclowes commented on web3/contract/utils.py:
```diff
@@ -60,10 +66,50 @@
 
 
+@curry
+def format_contract_call_return_data_curried(
```
I didn't think about this too hard, but could we repurpose `call_contract_function` to have a `curry` decorator? It looks like the logic and inputs are fairly similar?

fselmo replied:
Hmm I'm not sure. It's really only the `return_data` that we want to be able to pass in at a later time, everything else is preconfigured in the curried method. Since `call_contract_function` doesn't take the `return_data` argument I don't see how it can replace this. Is that kind of what you meant?

kclowes replied:
oh, gotcha. don't mind me :)

fselmo commented on web3/contract/utils.py:
```diff
@@ -108,6 +154,23 @@ def call_contract_function(
+                normalizers,
+                output_types,
+            ),
+            return_data[1][0],
```
make this more readable

fselmo commented on web3/contract/utils.py:
```diff
@@ -319,6 +382,33 @@ async def async_call_contract_function(
+            normalizers,
+            output_types,
+        )
+        if async_w3.provider.has_persistent_connection:
```
make this more readable pls

fselmo commented on web3/providers/ipc.py:
```diff
@@ -189,6 +192,47 @@ def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
                         timeout.sleep(0)
                         continue
 
+    def make_batch_request(
```
refactor similar logic to `make_request()` (on all providers where this isn't yet done)

fselmo commented on web3/method.py:
```diff
@@ -149,15 +153,29 @@ def __init__(
+        else:
+            return module.retrieve_caller_fn(self)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
```
~~Think about removing this?~~ [see below]

fselmo replied:
For clarity, this explicitly makes `Method` class callable... otherwise we get mypy conflicts, among other issues, such as:

```
web3/_utils/batching.py:128: error: "Method[Callable[..., Any]]" not callable  [operator]
```

The `batching.py` line in question:

```py
       for method, params in batch_payload.items():
            for param in params:
                self.add(method(param))  # <--- conflicting line
```

fselmo reviewed (commented):
From team code review 5/6/24

fselmo committed:
Clarify logic in contract call abstraction for batch requests

fselmo committed:
Refactor common ipc provider logic for making requests & batch requests.

fselmo committed:
More appropriately set batching logic on the JSONBaseProvider class

fselmo committed:
Attempt to mitigate issues in matching responses to request ids:

- Only guarantee response `id` order for batch requests if every
  response has an `id` field. Otherwise, rely on the provider
  to return the responses in order and hope for the best. An `id`
  should really ever only be missing in very rare cases where the
  request object itself was not able to be parsed correctly. So,
  hopefully this proves to be robust enough for most use cases.

kclowes commented on web3/providers/persistent/request_processor.py:
```diff
@@ -217,6 +232,28 @@ def get_request_information_for_response(
+            cached_request_info_for_id.response_formatters = (
+                compose(
+                    result_formatter,
+                    cached_request_info_for_id.response_formatters[0],
```
What do you think about making these `response_formatters` their own attributes instead of a tuple? Then we could reference like:
```suggestion
                    cached_request_info_for_id.result_formatters,
                    cached_request_info_for_id.error_formatters,
                    ...
```

fselmo replied:
Ah, yeah that could make this more readable for sure. I like this. I'll play around with that idea.

fselmo replied:
Since I think it's only in this part of the code base that unpacks them like this, what do you think about [unpacking the values](https://github.com/ethereum/web3.py/pull/3370/commits/56adab130945a29fd5f4d62990f54557c4792889) for better readability?

kclowes replied:
That works for me!

pacrob reviewed (approved):
I like it!

kclowes reviewed (approved):
This is awesome! Feel free to take or leave any comments I left. I wonder if we make the RequestsInformation object a little more robust we could avoid all of the indexing - [here](https://github.com/ethereum/web3.py/pull/3370/files#diff-c902d776bb4f676fa5cca4ee81273cd66d1113db8109d4f607167681b0b8bdeaR467) for example. I was thinking maybe have `RequestsInfo.method`, `RequestsInfo.params`, `RequestInfo.result_formatters`, `RequestInfo.error_formatters`, etc. We could either pass around the object, or just whatever attributes each function needs. If that doesn't make sense, feel free to push back :)

fselmo committed:
changes from comments on PR #3370:

- unpack response formatters for better readability

reedsa commented on web3/providers/persistent/request_processor.py:
```diff
@@ -95,7 +102,11 @@ def cache_request_information(
         method: RPCEndpoint,
         params: Any,
-        response_formatters: Tuple[Callable[..., Any], ...],
+        response_formatters: Tuple[
```
Might be good to create a new type for this, something like `ResponseFormatters`.

fselmo replied:
Yeah... these exist elsewhere in the code base and could honestly use some love throughout all of it. Separate issue to track maybe?

reedsa commented on web3/module.py:
```diff
@@ -55,9 +57,43 @@ def apply_result_formatters(
+    w3: Union["AsyncWeb3", "Web3"],
+    module: "Module",
+    method: Method[Callable[..., Any]],
+) -> Union[
+    Callable[..., Tuple[Tuple[RPCEndpoint, Any], Sequence[Any]]],
+    Callable[..., Coroutine[Any, Any, Tuple[Tuple[RPCEndpoint, Any], Sequence[Any]]]],
+]:
```
Maybe worth creating a type for `RequestInformation` and use that inside the `Callable[...]`s. There are a few places these are used. It could end up something like:
```suggestion
) -> Union[
    Callable[..., RequestInformation],
    Callable[..., Coroutine[Any, Any, RequestInformation],
]:
```

reedsa reviewed (approved):
Minor suggestions for typing but lgtm!
